File: src/main/java/com/zerone/core/Event.java
package com.zerone.core;

import java.util.ArrayList;

public class Event<T extends Object> {

    private final ArrayList<Subscriber<T>> subscriberList = new ArrayList<>();

    public void addSubscriber(Subscriber<T> subscriber) {
        subscriberList.add(subscriber);
    }

    public void removeSubscriber(Subscriber<T> subscriber) {
        subscriberList.remove(subscriber);
    }

    public void invoke(T param) {
        int size = subscriberList.size();
        for (int i = 0; i < size; i++) {
            subscriberList.get(i).onInvoked(param);
        }
    }

    public interface Subscriber<T> {
        void onInvoked(T param);
    }
}

File: src/main/kotlin/com/zerone/core/Event.kt
package com.zerone.core

import java.util.*

class Event<T : Any?> {

    private val subscriberList = ArrayList<Subscriber<T>>()

    fun addSubscriber(subscriber: Subscriber<T>) {
        subscriberList.add(subscriber)
    }

    fun removeSubscriber(subscriber: Subscriber<T>) {
        subscriberList.remove(subscriber)
    }

    operator fun invoke(param: T) {
        subscriberList.forEach { it.onEvent(param) }
    }

    interface Subscriber<T> {
        fun onEvent(param: T)
    }
}
